File: akka-docs/src/test/java/jdocs/stream/operators/source/Tick.java
/*
 * Copyright (C) 2009-2020 Lightbend Inc. <https://www.lightbend.com>
 */

package jdocs.stream.operators.source;

import akka.NotUsed;
import akka.actor.Cancellable;
import akka.actor.typed.ActorRef;
import akka.actor.typed.ActorSystem;
import akka.actor.typed.javadsl.AskPattern;
import java.util.concurrent.CompletionStage;
import java.time.Duration;

import akka.japi.Pair;
import akka.stream.javadsl.Flow;
import akka.stream.javadsl.Source;

public class Tick {

  // not really a runnable example, these are just pretend
  private ActorSystem<Void> system = null;
  private ActorRef<MyActor.Command> myActor = null;

  static class MyActor {
    interface Command {}

    static class Query implements Command {
      public final ActorRef<Response> replyTo;

      public Query(ActorRef<Response> replyTo) {
        this.replyTo = replyTo;
      }
    }

    static class Response {
      public final String text;

      public Response(String text) {
        this.text = text;
      }
    }
  }

  void simple() {
    // #simple
    Source.tick(
            Duration.ofSeconds(1), // delay of first tick
            Duration.ofSeconds(1), // delay of subsequent ticks
            "tick" // element emitted each tick
            )
        .runForeach(System.out::println, system);
    // #simple
  }

  void pollSomething() {
    // #poll-actor
    Source<String, Cancellable> periodicActorResponse =
        Source.tick(Duration.ofSeconds(1), Duration.ofSeconds(1), "tick")
            .mapAsync(
                1,
                notUsed -> {
                  CompletionStage<MyActor.Response> response =
                      AskPattern.ask(
                          myActor, MyActor.Query::new, Duration.ofSeconds(3), system.scheduler());
                  return response;
                })
            .map(response -> response.text);
    // #poll-actor

    // #zip-latest
    Flow<Integer, Pair<Integer, String>, NotUsed> zipWithLatestResponse =
        Flow.of(Integer.class).zipLatest(periodicActorResponse);
    // #zip-latest
  }
}

File: akka-docs/src/test/scala/docs/stream/operators/source/Tick.scala
/*
 * Copyright (C) 2009-2023 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.stream.operators.source

import akka.NotUsed
import akka.actor.Cancellable
import akka.actor.typed.ActorRef
import akka.actor.typed.ActorSystem
import akka.stream.scaladsl.Source
import akka.actor.typed.scaladsl.AskPattern._
import akka.stream.scaladsl.Flow
import akka.util.Timeout

import scala.annotation.nowarn
import scala.concurrent.Future
import scala.concurrent.duration._

@nowarn("msg=never used") // sample snippets
object Tick {
  // not really a runnable example, these are just pretend
  implicit val system: ActorSystem[Nothing] = null
  val myActor: ActorRef[MyActor.Command] = null;

  object MyActor {
    sealed trait Command {}
    case class Query(replyTo: ActorRef[Response]) extends Command
    case class Response(text: String)
  }

  def simple(): Unit = {
    // #simple
    Source
      .tick(
        1.second, // delay of first tick
        1.second, // delay of subsequent ticks
        "tick" // element emitted each tick
      )
      .runForeach(println)
    // #simple
  }

  def pollSomething(): Unit = {
    // #poll-actor
    val periodicActorResponse: Source[String, Cancellable] = Source
      .tick(1.second, 1.second, "tick")
      .mapAsync(1) { _ =>
        implicit val timeout: Timeout = 3.seconds
        val response: Future[MyActor.Response] = myActor.ask(MyActor.Query(_))
        response
      }
      .map(_.text);
    // #poll-actor

    // #zip-latest
    val zipWithLatestResponse: Flow[Int, (Int, String), NotUsed] =
      Flow[Int].zipLatest(periodicActorResponse);
    // #zip-latest
  }
}
